modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java

```diff
@@ -0,0 +1,143 @@
+                index("test", "test", String.valueOf(i * 10 + j), "field", j);
+            }
+        }
+        refresh();
+        assertHitCount(client().prepareSearch("test").setSize(0).get(), docs * threads.length);
+
+        final CountDownLatch start = new CountDownLatch(1);
+        final AtomicReference<Throwable> exceptionHolder = new AtomicReference<>();
+
+        for (int i = 0; i < threads.length; i++) {
+            final int threadNum = i;
+            assertHitCount(client().prepareSearch("test").setSize(0).setQuery(QueryBuilders.termQuery("field", threadNum)).get(), docs);
+
+            Runnable r = () -> {
+                try {
+                    start.await();
+
+                    assertThat(deleteByQuery().source("_all").filter(termQuery("field", threadNum)).refresh(true).get(),
+                            matcher().deleted(docs));
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                } catch (Throwable e) {
+                    exceptionHolder.set(e);
+                    Thread.currentThread().interrupt();
+                }
+            };
+            threads[i] = new Thread(r);
+            threads[i].start();
+        }
+
+        start.countDown();
+        for (Thread thread : threads) {
+            thread.join();
+        }
+
+        Throwable assertionError = exceptionHolder.get();
+        if (assertionError != null) {
+            throw assertionError;
+        }
+
+        for (int i = 0; i < threads.length; i++) {
+            assertHitCount(client().prepareSearch("test").setSize(0).setQuery(QueryBuilders.termQuery("field", i)).get(), 0);
+        }
+    }
+
+    public void testConcurrentDeleteByQueriesOnSameDocs() throws Throwable {
+        assertAcked(prepareCreate("test").setSettings(Settings.builder().put("index.refresh_interval", -1)));
+        ensureGreen();
+
+        final long docs = randomIntBetween(50, 100);
+        for (int i = 0; i < docs; i++) {
+            index("test", "test", String.valueOf(i), "foo", "bar");
+        }
+        refresh();
+        assertHitCount(client().prepareSearch("test").setSize(0).get(), docs);
+
+        final Thread[] threads =  new Thread[scaledRandomIntBetween(2, 9)];
+
+        final CountDownLatch start = new CountDownLatch(1);
+        final AtomicReference<Throwable> exceptionHolder = new AtomicReference<>();
+
+        final MatchQueryBuilder query = matchQuery("foo", "bar");
+        final AtomicLong deleted = new AtomicLong(0);
+
+        for (int i = 0; i < threads.length; i++) {
+            assertHitCount(client().prepareSearch("test").setSize(0).setQuery(query).get(), docs);
+
+            Runnable r = () -> {
+                try {
+                    start.await();
+
+                    BulkIndexByScrollResponse response = deleteByQuery().source("_all").filter(query).refresh(true).get();
+                    deleted.addAndGet(response.getDeleted());
```

Comment:
You might want to comment that you expect some of these requests to fail with a version conflict but that the total will be the total number of docs. Right?